## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [328]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [329]:
# View the column names in the data
df_cols=df.columns
df_cols

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [330]:
# Use the describe function to gather some basic statistics
df_stats=df.describe()
df_stats

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [334]:
# Use this space to do any additional research
# and familiarize yourself with the data.

#print(df.isnull().sum())  # are there an nulls
print("-"*300)
#print(df.dtypes)  #data types of the columns
df["year_week"]=df["order_year"].astype(str) + "_" + df["order_week"].astype(str)
print("-"*300)
print(df["year_week"].unique()) #see the whole range of weeks and years





------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
['2023_17' '2023_20' '2023_4' '2023_21' '2023_18' '2023_16' '2023_7'
 '2023_19' '2023_15' '2023_6' '2023_14' '2023_22' '2023_10' '2023_11'
 '2023_3' '2023_2' '2023_9' '2023_13' '2023_5' '2023_1' '2023_12' '2023_8'
 '2022_52']


In [335]:
# What three item categories had the most entries?
print("categories: ",df["category"].unique())  # number of categories
print("-"*300)
df["category"].value_counts(ascending= False).head(3) #top 3


categories:  ['decor' 'consumables' 'software' 'furniture' 'electronics']
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [336]:
# For the category with the most entries, which subcategory had the most entries?
#reduce to rows within top 3 categories
df_top3_cat=df[df["category"].isin(["consumables","furniture","software"])]

#count the entries of subcategory within top 3 categories.  head(1) finds the most
df_top3_cat["subcategory"].value_counts(ascending=False).head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [337]:
# Which five clients had the most entries in the data?
df["client_id"].value_counts(ascending= False).head(5) #top 5


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [339]:
# Store the client ids of those top 5 clients in a list.

#.index.tolist() from google
top5_clients_list=df["client_id"].value_counts(ascending= False).head(5).index.tolist() 

top5_clients_list



[33615, 66037, 46820, 38378, 24741]

In [343]:
# How many total units (the qty column) did the client with the most entries order order?

#from a couple of steps earlier we know client ID 33615 has the most entries
df[df["client_id"]==33615]["qty"].sum()

#or alternatively using the list of clients
#df[df["client_id"]==top5_clients_list[0]]["qty"].sum()



64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [344]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df_with_cost=(df
              .pipe(lambda x: x.assign(all_item_price=x["qty"]*x["unit_price"])))

#look at the data
df_with_cost.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number year_week  \
0     1096.80     762.71         7.50  105            1   2023_17   
1       24.95      15.09         1.49   21            0   2023_20   
2       13.52       7.86         1.68   39            6    2023_4   
3       36.42      24.85         1.23   29            3   2023_17   
4      195.10     108.17        46.43   20            1   2023_21   

   all_item_price  
0       115164.00  
1          523.95  
2          527.28  
3         1056.18  
4         3902.00

In [345]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.

#create shipping_unit_price function
def weight_unit_price(weight):
    if(weight<=50):
        weight_unit_price=10.00
    else:
        weight_unit_price=7.00
    return weight_unit_price

#continuing with the pipe syntax
df_with_cost=(df
              #from previous step
              .pipe(lambda x: x.assign(all_item_price=x["qty"]*x["unit_price"])) 
              # new to this step
              .pipe(lambda x: x.assign(total_weight=x["qty"]*x["unit_weight"]))
              .pipe(lambda x: x.assign(shipping_unit_price=x["total_weight"].apply(weight_unit_price)))
              .pipe(lambda x: x.assign(shipping_price=x["shipping_unit_price"]*x["total_weight"]))
)

#look at the data
df_with_cost.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number  year_week  all_item_price  total_weight shipping_unit_price  \
0            1    2023_17       115164.00        787.50                 7.0   
1            0    2023_20          523.95         31.29                10.0   
2            6     2023_4          527.28         65.52                 7.0   
3            3    2023_17         1056.18         35.67                10.0   
4            1    2023_21         3902.00        928.60                 7.0   

   shipping_price  
0         5512.50  
1          312.90  
2          458.64  
3          356.70  
4         6500.20  

[5 rows x 23 columns]

In [346]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%

#define function for total cost to make total cost column cleaner
def total_cost(base_cost, shipping_cost,sales_tax):
    cost_no_tax=base_cost+shipping_cost
    cost_of_tax=cost_no_tax*sales_tax
    total_cost=cost_no_tax+cost_of_tax
    return total_cost

#continuing with the pipe syntax
df_with_cost=(df
              #from previous steps
              .pipe(lambda x: x.assign(all_item_price=x["qty"]*x["unit_price"])) 
              .pipe(lambda x: x.assign(total_weight=x["qty"]*x["unit_weight"]))
              .pipe(lambda x: x.assign(shipping_unit_price=x["total_weight"].apply(weight_unit_price)))
              .pipe(lambda x: x.assign(shipping_price=x["shipping_unit_price"]*x["total_weight"]))
              #new to this step
              .pipe(lambda x: x.assign(sales_tax=.0925))
              #needed bing copilot to send multiple arguments but did not want to create intermediate columns
              .pipe(lambda x: x.assign(total_price=x.apply(lambda row: 
              total_cost(row["all_item_price"],row["shipping_price"],row["sales_tax"]),axis=1)))
              )


#look at the data
df_with_cost.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_weight  qty line_number  year_week  \
0          17        2023  ...        7.50  105           1    2023_17   
1          20        2023  ...        1.49   21           0    2023_20   
2           4        2023  ...        1.68   39           6     2023_4   
3          17        2023  ...        1.23   29           3    2023_17   
4          21        2023  ...       46.43   20           1    2023_21   

   all_item_price  total_weight  shipping_unit_price  shipping_price  \
0       115164.00        787.50                  7.0         5512.50   
1          523.95         31.29                 10.0          312.90   
2          527.28         65.52                  7.0          458.64   
3         1056.18         35.67                 10.0          356.70   
4         3902.00        928.60                  7.0         6500.20   

  sales_tax    total_price  
0    0.0925  131839.076250  
1    0.0925     914.258625  
2    0.0925    1077.117600  
3    0.0925    1543.571400  
4    0.0925   11364.403500  

[5 rows x 25 columns]

In [283]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).

#continuing with the pipe syntax
df_with_cost=(df
              #from previous steps
              .pipe(lambda x: x.assign(all_item_price=x["qty"]*x["unit_price"])) 
              .pipe(lambda x: x.assign(total_weight=x["qty"]*x["unit_weight"]))
              .pipe(lambda x: x.assign(shipping_unit_price=x["total_weight"].apply(weight_unit_price)))
              .pipe(lambda x: x.assign(shipping_price=x["shipping_unit_price"]*x["total_weight"]))
              .pipe(lambda x: x.assign(sales_tax=.0925))
              #needed bing copilot to send multiple arguments but did not want to create intermediate columns
              .pipe(lambda x: x.assign(total_price=x.apply(lambda row: 
              total_cost(row["all_item_price"],row["shipping_price"],row["sales_tax"]),axis=1)))
            #new to this step
              .pipe(lambda x: x.assign(all_item_cost=x["qty"]*x["unit_cost"]))
              .pipe(lambda x: x.assign(line_cost=x.apply(lambda row: 
              total_cost(row["all_item_cost"],row["shipping_price"],0),axis=1)))
              )

#look at the data
df_with_cost.head()




first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...  qty line_number all_item_price  total_weight  \
0          17        2023  ...  105           1      115164.00        787.50   
1          20        2023  ...   21           0         523.95         31.29   
2           4        2023  ...   39           6         527.28         65.52   
3          17        2023  ...   29           3        1056.18         35.67   
4          21        2023  ...   20           1        3902.00        928.60   

   shipping_unit_price  shipping_price  sales_tax    total_price  \
0                  7.0         5512.50     0.0925  131839.076250   
1                 10.0          312.90     0.0925     914.258625   
2                  7.0          458.64     0.0925    1077.117600   
3                 10.0          356.70     0.0925    1543.571400   
4                  7.0         6500.20     0.0925   11364.403500   

   all_item_cost  line_cost  
0       80084.55   85597.05  
1         316.89     629.79  
2         306.54     765.18  
3         720.65    1077.35  
4        2163.40    8663.60  

[5 rows x 26 columns]

In [380]:
# Create a column for the profit of each line using line cost and line price
#continuing with the pipe syntax

#define a function to get 2 decimal places
def format_cost(value):
        return round(value,2)

df_with_cost=(df
              #from previous steps
              .pipe(lambda x: x.assign(all_item_price=x["qty"]*x["unit_price"])) 
              .pipe(lambda x: x.assign(total_weight=x["qty"]*x["unit_weight"]))
              .pipe(lambda x: x.assign(shipping_unit_price=x["total_weight"].apply(weight_unit_price)))
              .pipe(lambda x: x.assign(shipping_price=x["shipping_unit_price"]*x["total_weight"]))
              .pipe(lambda x: x.assign(sales_tax=.0925))
              #needed bing copilot to send multiple arguments but did not want to create intermediate columns
              .pipe(lambda x: x.assign(total_price=x.apply(lambda row: 
              total_cost(row["all_item_price"],row["shipping_price"],row["sales_tax"]),axis=1)))
              .pipe(lambda x: x.assign(all_item_cost=x["qty"]*x["unit_cost"]))
              .pipe(lambda x: x.assign(line_cost=x.apply(lambda row: 
              total_cost(row["all_item_cost"],row["shipping_price"],0),axis=1)))
              #new to this step
              .pipe(lambda x: x.assign(profit=x["total_price"]-x["line_cost"]))
              )

#formatting columns
columns_to_format=["all_item_price","shipping_unit_price","shipping_price","total_price","all_item_cost","line_cost","profit"]
df_with_cost_formatted=df_with_cost
df_with_cost_formatted[columns_to_format]=df_with_cost[columns_to_format].map(format_cost)

#look at the data
df_with_cost_formatted.head()





first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... year_week all_item_price total_weight  \
0          17        2023  ...   2023_17      115164.00       787.50   
1          20        2023  ...   2023_20         523.95        31.29   
2           4        2023  ...    2023_4         527.28        65.52   
3          17        2023  ...   2023_17        1056.18        35.67   
4          21        2023  ...   2023_21        3902.00       928.60   

   shipping_unit_price  shipping_price  sales_tax  total_price  all_item_cost  \
0                  7.0         5512.50     0.0925    131839.08       80084.55   
1                 10.0          312.90     0.0925       914.26         316.89   
2                  7.0          458.64     0.0925      1077.12         306.54   
3                 10.0          356.70     0.0925      1543.57         720.65   
4                  7.0         6500.20     0.0925     11364.40        2163.40   

  line_cost    profit  
0  85597.05  46242.03  
1    629.79    284.47  
2    765.18    311.94  
3   1077.35    466.22  
4   8663.60   2700.80  

[5 rows x 28 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [381]:
# Check your work using the totals above
#just get the columns used for the check
check=df_with_cost_formatted.loc[:,["order_id","total_price"]]
#limit rows to order ids to the receipts in hand
check=check[check["order_id"].isin([2742071,2173913,6128929])]
#sum up for each order id since there are multiple items per order
check=check.groupby("order_id")["total_price"].sum().reset_index()

check.head()

#print(df_with_cost_formatted["total_price"].dtype)

order_id  total_price
0   2173913    162388.71
1   2742071    152811.89
2   6128929    923441.25

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [421]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.

#limit clients to top 5 list and then sum the total price for each
top5_spend=(df_with_cost_formatted
            .pipe(lambda df: df[df["client_id"].isin(top5_clients_list)])
            .pipe(lambda df: df.groupby("client_id")["total_price"].sum().      reset_index())
            #needed some chatgpt help with getting the sort in same order as the client list
            .pipe(lambda df: df.sort_values(by="client_id", key=lambda x: pd.Categorical(x, categories=top5_clients_list,ordered=True)))           
            )

top5_spend

#the line below does work but still not in the order of top5_clients_list.
#df_with_cost_formatted[df_with_cost_formatted["client_id"].isin(top5_clients_list)].groupby("client_id")["total_price"].sum()



client_id  total_price
1      33615   8377308.52
4      66037  10259514.80
3      46820   9743794.33
2      38378  12906550.87
0      24741  82268892.04

In [478]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit.
 
top5_spend=(df_with_cost_formatted
            .pipe(lambda df: df[["client_id","qty","shipping_price","total_price","profit"]])
            .pipe(lambda df: df[df["client_id"].isin(top5_clients_list)])
            .pipe(lambda df: df.groupby("client_id").sum().reset_index())
            #needed some chatgpt help with getting the sort in same order as the client list
            .pipe(lambda df: df.sort_values(by="client_id", key=lambda x: pd.Categorical(x, categories=top5_clients_list,ordered=True)))           
            )

top5_spend


client_id     qty  shipping_price  total_price       profit
1      33615   64313      1828984.89   8377308.52   2201994.61
4      66037   43018      1395151.85  10259514.80   3255031.82
3      46820   75768      1601448.84   9743794.33   2736602.70
2      38378   73667      3429455.40  12906550.87   3271829.89
0      24741  239862      5126448.37  82268892.04  36579992.33

In [511]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 
money_cols=["shipping_price","total_price","profit"]
money_cols

# Define a function that converts a dollar amount to millions.
def currency_format_millions(col):
    col_millions=col/1000000
    return col_millions

# Apply the currency_format_millions function to only the money columns.
top5_spend_pres=top5_spend.copy()
top5_spend_pres[money_cols]=top5_spend_pres[money_cols].apply(currency_format_millions)

# Rename the columns to reflect the change in the money format. 
top5_spend_pres.rename(columns={"client_id":"Client ID", "qty":"Units","shipping_price":"Shipping (millions)","total_price":"Total Revenue (millions)","profit":"Total Profit (millions)"})

#make multi row columns names
top5_spend_pres.columns=pd.MultiIndex.from_tuples([
    ("Client","ID"),
    ("Units",""),
    ("Shipping", "(millions)"),
    ("Total Revenue", "(millions)"),
    ("Total Profit", "(millions)")
])

top5_spend_pres


Client   Units   Shipping Total Revenue Total Profit
      ID         (millions)    (millions)   (millions)
1  33615   64313   1.828985      8.377309     2.201995
4  66037   43018   1.395152     10.259515     3.255032
3  46820   75768   1.601449      9.743794     2.736603
2  38378   73667   3.429455     12.906551     3.271830
0  24741  239862   5.126448     82.268892    36.579992

In [525]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.

#top5_spend_pres_sort=top5_spend_pres.sort_values
top5_spend_pres_sort=top5_spend_pres.sort_values(('Total Profit','(millions)'),ascending=False)

top5_spend_pres_sort



Client   Units   Shipping Total Revenue Total Profit
      ID         (millions)    (millions)   (millions)
0  24741  239862   5.126448     82.268892    36.579992
2  38378   73667   3.429455     12.906551     3.271830
4  66037   43018   1.395152     10.259515     3.255032
3  46820   75768   1.601449      9.743794     2.736603
1  33615   64313   1.828985      8.377309     2.201995

In [526]:
print("Assuming the most profit is coming from the clients who have the most orders, client 24741 was our biggest customer between week 52 2022 and week 22 2023.  We generated $36.6 million dollars from client 24741.  This was more than 10 times the next highest profit client.  Of course all clients are important, but client 24741 bought a particularly large number of unit and we should quickly address any concerns they may have in order to secure future business.")

Assuming the most profit is coming from the clients who have the most orders, client 24741 was our biggest customer between week 52 2022 and week 22 2023.  We generated $36.6 million dollars from client 24741.  This was more than 10 times the next highest profit client.  Of course all clients are important, but client 24741 bought a particularly large number of unit and we should quickly address any concerns they may have in order to secure future business.
